[WIP] Implement the importing of optics

lamemakes opened the pull request:
Closes #148. This is a rough first pass at handling the importing of blocked/liked/disliked sites via `.optic` files from previous exports or other means. 

**Changes TLDR:**
- Allow for the "uploading" of `.optic` files (all optic file parsing is client side)
- Fixed the way the server exports optic files to ensure the expected formatting is received (axum was trying to export JSON, causing escaped quotes and other weird stuff)
- Replaced the `Ranked` type with an enum to *(hopefully?)* simplify some readability & programmatic things now/down the line. If this is undesired it can totally be flipped back. 
- Once extracted & imported, sites are appended to the existing store as suggested to avoid weird overwrites
- Fixed a typo. `SiteRakings` -> `SiteRankings`

**Screenshots:**
![Screenshot from 2024-02-22 23-53-32](https://github.com/StractOrg/stract/assets/83597346/2e7fda76-0296-4517-b8c3-cd06b0b5b293)
*disclaimer: no beef or love for any of these websites, first ones to come up searches*


Needs more cleaning up, commenting and to be a bit more robust but wanted to snag some early feedback before committing to it more. Are y'all doing JSDoc? I just did inline comments for now but planned to go back with docstrings.

Cheers!

lamemakes commented:
That failure is odd, I can't seem to recreate it locally.

mikkeldenker commented on frontend/src/lib/rankings.ts:
```diff
@@ -29,6 +35,35 @@ export const rankedToRankings = (ranked: RankedSites): SiteRakings => {
   return result;
 };
 
+// Map regexs to their respective rankings
+const reToRankMap = {
+  [Ranking.LIKED]: /Like\(\s*Site\s*\(\s*"(\S*)"\s*\)\s*\);/g,
+  [Ranking.DISLIKED]: /Dislike\(\s*Site\s*\(\s*"(\S*)"\s*\)\s*\);/g,
+  [Ranking.BLOCKED]: /Rule\s*{\s*Matches\s*{\s*Site\("\|(\S*)\|"\),\s*},\s*Action\(Discard\)\s*};/g
+}
+
+export const importedOpticToRankings = (optic: string): SiteRankings => {
+  const rankedSites: RankedSites = {
+    liked: [],
+    disliked: [],
+    blocked: [],
+  };
+
+  for(const [_, ranking] of Object.entries(Ranking)){
+    let re = reToRankMap[ranking]
+    let match: RegExpExecArray | null
+
+    do {
+      match = re.exec(optic);
+      if (match) {
+        rankedSites[ranking].push(match[1])
+      }
+    } while (match)
+  }
+
+  return rankedToRankings(rankedSites)
+}
+
```
I think we should keep all optics parsing inside the `optics` crate in case we want to make changes to the syntax later.
The most elegant solution would probably be to create a wasm crate that exposes this functionality to the frontend so the parsing still happens client-side. Alternatively we can create a `/beta/api/hosts/import` endpoint for the time being and then move the functionality to a wasm create later.

lamemakes replied:
Totally makes sense. To me WASM makes the most sense to preserve some of your resources on the server side. Been looking for an excuse to play with it for a while anyway. Some optic validation would be nice too, is that the goal with #147?

mikkeldenker commented on frontend/src/routes/settings/sites/+page.svelte:
```diff
@@ -74,5 +114,7 @@
     {#each buttons as { text, clear }}
       <Button on:click={clearAndExport({ clear })}>{text}</Button>
     {/each}
+    <input type="file" accept=".optic" id="optic-import" multiple hidden/>
+    <Button on:click={importOpticFile}>Import from optic</Button>
```
I think a `<label><input type="file" ... /></label>` structure where we style the label with the same styles as the button component would be preferred from an accessibility perspective. We can then extract line `46..=64` into a separate function and have it as an `on:change` attribute on the input element.

lamemakes replied:
Also agreed, there was a lot of beautiful button styling logic happening in that component & I was trying to minimize duplicate code which is why I took the more roundabout way to things. Could that logic be extracted to a function in some kind of utils file? or did you have other ideas?

mikkeldenker reviewed (changes requested):
Thank you, this is a great first pass! I fully agree the `Ranked` enum is more readable, and nice catch with the axum return type.

I think there is a weird non-deterministic bug in the test that failed in the CI. I re-ran it and everything seems to be fine. I'll look into it a bit more thoroughly next week and see if I can track it down.

lamemakes commented:
Alright, new updates and I think I covered most of my bases. wasm was a great call, made life so much easier in terms of optic parsing.

**New TLDR:**
 - Added a crate to be used for general client-side WASM bindings. I figured this would add less complexity if any other client side features were added, as a new struct can just be created in that crate. Let me know if you'd rather them broken out. This crate houses the `Optics` struct that contains the associated function `parse_preference_optic`.
 - Additions to vite + svelte configs to allow for packaging & execution of WASM code
 - Extraction of button tailwind logic to `themes.ts` to allow for reuse by labels and other elements
 - Updated `CONTRIBUTING.md` to reflect the need for [wasm-pack](https://rustwasm.github.io/wasm-pack/installer/)

Let me know what you think! Only issue I saw was there's an issue when it comes to importing the `blocked` sites. I think that maybe stems from [here](https://github.com/StractOrg/stract/blob/main/crates/optics/src/lib.rs#L102), as it comes back as an empty array. I didn't really have the historical context to know what `$discard` meant in that comment.

lamemakes commented:
also kept the `Cargo.lock` as I considered the generated WASM a binary, can toss it if you'd prefer though.

lamemakes commented:
gahhh this killed me tonight. Running the CI scripts locally all is passing. Throwing in the towel for now - the only issue that's left is that the [CSP typing](https://github.com/sveltejs/kit/blob/41bb2c0fc752843651e00de133da618884c2c5ee/packages/kit/types/index.d.ts#L1420) doesn't allow [`['*']`](https://github.com/StractOrg/stract/blame/main/frontend/svelte.config.js#L34). I can't typecast either because it's a JS file which is so weird to me. All should pass outside of that.

The wasm-pack has this awful build log though ([that's pretty well complained about](https://github.com/rustwasm/wasm-bindgen/issues/2547)) that I can't seem to suppress either.

lamemakes commented:
Thank you so much for the assist! I completely missed #109! That would've been a fantastic resource if I had found it. did you see my comment regarding the blocked sites? Not sure if you had a different experience:

> Only issue I saw was there's an issue when it comes to importing the `blocked` sites. I think that maybe stems from [here](https://github.com/StractOrg/stract/blob/main/crates/optics/src/lib.rs#L102), as it comes back as an empty array. I didn't really have the historical context to know what `$discard` meant in that comment.

I can write-up an issue if needed. Cheers!

mikkeldenker commented:
Oh, I totally forgot about that! You are completely right.
The comment in the code is way out of date syntax wise, but this is indeed what's causing this issue. The problem essentially is that blocked sites are stored in the optic as `Rule { Matches { Site("|...|") }, Action(Discard) }` rules and these simply get parsed directly into the `rules` vec so I thought there was no need to add them to the `HostRankings` as the rules would already block the sites.

I think the best solution is to add a special check for these rules when `RawOptic` is converted to `Optic` and simply add these sites to the `blocked` vec instead. They would still generate exactly the same tantivy queries as before and still correctly block the sites, but this would allow us to extract the sites during import. It would probably also make it easier to merge optics, so I think it's just a better approach all-around.

No need to open an issue, I'll fix it right away.
